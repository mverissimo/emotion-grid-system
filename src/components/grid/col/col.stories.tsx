import type { Meta, StoryObj } from '@storybook/react-vite';

import Col from './col';

const meta = {
  component: Col,
  title: 'Components/Grid/Col',
} satisfies Meta<typeof Col>;

export const Base: StoryObj<typeof meta> = {
  args: {
    size: {
      md: 1,
    },
  },
};

export default meta;
