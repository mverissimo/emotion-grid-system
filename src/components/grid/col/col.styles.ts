import { css, cx } from '@emotion/css';
import type { Theme } from '@emotion/react';

import { mediaOnly, responsive } from '../../../utils/media';
import type { ColProps } from './col';

const initial = (theme: Theme) =>
  cx(
    css({
      label: 'grid-col',

      flex: '1 0 0%',
      boxSizing: 'border-box',

      width: '100%',
      maxWidth: '100%',
    }),
    responsive(theme, (breakpoint) => ({
      paddingLeft: `calc(${theme.grid?.gutter[breakpoint]} / 2)`,
      paddingRight: `calc(${theme.grid?.gutter[breakpoint]} / 2)`,
    })),
  );

const size = (theme: Theme, size: ColProps['size']) => {
  if (size) {
    return cx(
      css({
        label: 'grid-col--size',
      }),
      mediaOnly(theme, size, (breakpoint) => ({
        flex: `${size[breakpoint] === 'auto' ? 'auto' : `0 0 calc((${size[breakpoint]} / ${theme.grid?.columns[breakpoint]}) * 100%)`}`,
      })),
    );
  }
};

const offset = (theme: Theme, offset: ColProps['offset']) => {
  if (offset) {
    return cx(
      css({
        label: 'grid-col--offset',
      }),
      mediaOnly(theme, offset, (breakpoint) => ({
        marginLeft: `calc((${offset[breakpoint]} / ${theme.grid?.columns[breakpoint]}) * 100%)`,
      })),
    );
  }
};

export default {
  initial,
  size,
  offset,
};
