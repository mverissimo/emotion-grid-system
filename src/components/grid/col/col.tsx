import type { HTMLAttributes, Ref } from 'react';

import { cx } from '@emotion/css';
import type { Breakpoint } from '../../../theme';

import { useTheme } from '../../../hooks/use-theme';

import styles from './col.styles';

export interface ColProps extends HTMLAttributes<HTMLDivElement> {
  /**
   * Set the column size
   */
  size?: Partial<Record<Breakpoint, number | 'auto'>>;

  /**
   * Set the column offset
   */
  offset?: Partial<Record<Breakpoint, number>>;
}

function Col(props: ColProps, ref?: Ref<HTMLDivElement>) {
  const { size, offset, className, ...rest } = props;

  const theme = useTheme();

  return (
    <div
      data-grid="col"
      className={cx(
        styles.initial(theme),
        styles.size(theme, size),
        styles.offset(theme, offset),
        className,
      )}
      ref={ref}
      {...rest}
    />
  );
}

export default Col;
