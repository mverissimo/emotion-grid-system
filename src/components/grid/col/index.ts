export { default as Col } from './col';
export type { ColProps } from './col';
