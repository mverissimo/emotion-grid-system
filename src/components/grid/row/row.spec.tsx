import { render } from '@testing-library/react';

import { Row } from '.';
import { Col } from '../col';

describe('Row', () => {
  it('should render with default styles', () => {
    const { container } = render(<Row>Row</Row>);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render with reverse styles when passed the reverse prop', () => {
    const { container } = render(
      <Row
        reverse={{
          sm: true,
          xl: true,
        }}
      >
        Row reverse
      </Row>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render align styles when passed the align prop', () => {
    const { container } = render(
      <Row
        align={{
          md: 'center',
        }}
      >
        Row
      </Row>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render justify styles when passed the justify prop', () => {
    const { container } = render(
      <Row
        justify={{
          xl: 'space-between',
        }}
      >
        Row justify
      </Row>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render with noGutters styles when passed the noGutters prop', () => {
    const { container } = render(
      <Row noGutters={true}>
        <Col>Col</Col>
      </Row>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
