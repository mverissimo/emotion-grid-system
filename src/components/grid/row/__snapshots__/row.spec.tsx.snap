// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Row > should render align styles when passed the align prop 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  box-sizing: border-box;
  margin-left: calc((2rem / 2) * -1);
  margin-right: calc((2rem / 2) * -1);
}

@media (min-width: 40rem) {
  .emotion-0 {
    margin-left: calc((2rem / 2) * -1);
    margin-right: calc((2rem / 2) * -1);
  }
}

@media (min-width: 48rem) {
  .emotion-0 {
    margin-left: calc((2rem / 2) * -1);
    margin-right: calc((2rem / 2) * -1);
  }
}

@media (min-width: 64rem) {
  .emotion-0 {
    margin-left: calc((2rem / 2) * -1);
    margin-right: calc((2rem / 2) * -1);
  }
}

@media (min-width: 80rem) {
  .emotion-0 {
    margin-left: calc((2rem / 2) * -1);
    margin-right: calc((2rem / 2) * -1);
  }
}

@media (min-width: 96rem) {
  .emotion-0 {
    margin-left: calc((2rem / 2) * -1);
    margin-right: calc((2rem / 2) * -1);
  }
}

@media (min-width: 48rem) {
  .emotion-0 {
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
  }
}

<div
  class="emotion-0"
  data-grid="row"
>
  Row
</div>
`;

exports[`Row > should render justify styles when passed the justify prop 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  box-sizing: border-box;
  margin-left: calc((2rem / 2) * -1);
  margin-right: calc((2rem / 2) * -1);
}

@media (min-width: 40rem) {
  .emotion-0 {
    margin-left: calc((2rem / 2) * -1);
    margin-right: calc((2rem / 2) * -1);
  }
}

@media (min-width: 48rem) {
  .emotion-0 {
    margin-left: calc((2rem / 2) * -1);
    margin-right: calc((2rem / 2) * -1);
  }
}

@media (min-width: 64rem) {
  .emotion-0 {
    margin-left: calc((2rem / 2) * -1);
    margin-right: calc((2rem / 2) * -1);
  }
}

@media (min-width: 80rem) {
  .emotion-0 {
    margin-left: calc((2rem / 2) * -1);
    margin-right: calc((2rem / 2) * -1);
  }
}

@media (min-width: 96rem) {
  .emotion-0 {
    margin-left: calc((2rem / 2) * -1);
    margin-right: calc((2rem / 2) * -1);
  }
}

@media (min-width: 80rem) {
  .emotion-0 {
    -webkit-box-pack: justify;
    -webkit-justify-content: space-between;
    justify-content: space-between;
  }
}

<div
  class="emotion-0"
  data-grid="row"
>
  Row justify
</div>
`;

exports[`Row > should render with default styles 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  box-sizing: border-box;
  margin-left: calc((2rem / 2) * -1);
  margin-right: calc((2rem / 2) * -1);
}

@media (min-width: 40rem) {
  .emotion-0 {
    margin-left: calc((2rem / 2) * -1);
    margin-right: calc((2rem / 2) * -1);
  }
}

@media (min-width: 48rem) {
  .emotion-0 {
    margin-left: calc((2rem / 2) * -1);
    margin-right: calc((2rem / 2) * -1);
  }
}

@media (min-width: 64rem) {
  .emotion-0 {
    margin-left: calc((2rem / 2) * -1);
    margin-right: calc((2rem / 2) * -1);
  }
}

@media (min-width: 80rem) {
  .emotion-0 {
    margin-left: calc((2rem / 2) * -1);
    margin-right: calc((2rem / 2) * -1);
  }
}

@media (min-width: 96rem) {
  .emotion-0 {
    margin-left: calc((2rem / 2) * -1);
    margin-right: calc((2rem / 2) * -1);
  }
}

<div
  class="emotion-0"
  data-grid="row"
>
  Row
</div>
`;

exports[`Row > should render with noGutters styles when passed the noGutters prop 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  box-sizing: border-box;
  margin-left: calc((2rem / 2) * -1);
  margin-right: calc((2rem / 2) * -1);
}

@media (min-width: 40rem) {
  .emotion-0 {
    margin-left: calc((2rem / 2) * -1);
    margin-right: calc((2rem / 2) * -1);
  }
}

@media (min-width: 48rem) {
  .emotion-0 {
    margin-left: calc((2rem / 2) * -1);
    margin-right: calc((2rem / 2) * -1);
  }
}

@media (min-width: 64rem) {
  .emotion-0 {
    margin-left: calc((2rem / 2) * -1);
    margin-right: calc((2rem / 2) * -1);
  }
}

@media (min-width: 80rem) {
  .emotion-0 {
    margin-left: calc((2rem / 2) * -1);
    margin-right: calc((2rem / 2) * -1);
  }
}

@media (min-width: 96rem) {
  .emotion-0 {
    margin-left: calc((2rem / 2) * -1);
    margin-right: calc((2rem / 2) * -1);
  }
}

.emotion-0 [data-grid=col] {
  padding-left: 0!important;
  padding-right: 0!important;
}

.emotion-1 {
  -webkit-flex: 1 0 0%;
  -ms-flex: 1 0 0%;
  flex: 1 0 0%;
  box-sizing: border-box;
  width: 100%;
  max-width: 100%;
  padding-left: calc(2rem / 2);
  padding-right: calc(2rem / 2);
}

@media (min-width: 40rem) {
  .emotion-1 {
    padding-left: calc(2rem / 2);
    padding-right: calc(2rem / 2);
  }
}

@media (min-width: 48rem) {
  .emotion-1 {
    padding-left: calc(2rem / 2);
    padding-right: calc(2rem / 2);
  }
}

@media (min-width: 64rem) {
  .emotion-1 {
    padding-left: calc(2rem / 2);
    padding-right: calc(2rem / 2);
  }
}

@media (min-width: 80rem) {
  .emotion-1 {
    padding-left: calc(2rem / 2);
    padding-right: calc(2rem / 2);
  }
}

@media (min-width: 96rem) {
  .emotion-1 {
    padding-left: calc(2rem / 2);
    padding-right: calc(2rem / 2);
  }
}

<div
  class="emotion-0"
  data-grid="row"
>
  <div
    class="emotion-1"
    data-grid="col"
  >
    Col
  </div>
</div>
`;

exports[`Row > should render with reverse styles when passed the reverse prop 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  box-sizing: border-box;
  margin-left: calc((2rem / 2) * -1);
  margin-right: calc((2rem / 2) * -1);
}

@media (min-width: 40rem) {
  .emotion-0 {
    margin-left: calc((2rem / 2) * -1);
    margin-right: calc((2rem / 2) * -1);
  }
}

@media (min-width: 48rem) {
  .emotion-0 {
    margin-left: calc((2rem / 2) * -1);
    margin-right: calc((2rem / 2) * -1);
  }
}

@media (min-width: 64rem) {
  .emotion-0 {
    margin-left: calc((2rem / 2) * -1);
    margin-right: calc((2rem / 2) * -1);
  }
}

@media (min-width: 80rem) {
  .emotion-0 {
    margin-left: calc((2rem / 2) * -1);
    margin-right: calc((2rem / 2) * -1);
  }
}

@media (min-width: 96rem) {
  .emotion-0 {
    margin-left: calc((2rem / 2) * -1);
    margin-right: calc((2rem / 2) * -1);
  }
}

@media (min-width: 40rem) {
  .emotion-0 {
    -webkit-flex-direction: row-reverse;
    -ms-flex-direction: row-reverse;
    flex-direction: row-reverse;
  }
}

@media (min-width: 80rem) {
  .emotion-0 {
    -webkit-flex-direction: row-reverse;
    -ms-flex-direction: row-reverse;
    flex-direction: row-reverse;
  }
}

<div
  class="emotion-0"
  data-grid="row"
>
  Row reverse
</div>
`;
